File: backend/utils/validation.py
"""Shared Portuguese customer-data validation helpers."""

from __future__ import annotations

import re
from typing import Optional

DISPOSABLE_EMAIL_DOMAINS = {
    "10minutemail.com",
    "guerrillamail.com",
    "mailinator.com",
    "tempmail.com",
    "temp-mail.org",
    "yopmail.com",
}

EMAIL_RE = re.compile(r"^[^\s@<>]+@[^\s@<>]+\.[^\s@<>]{2,}$")
NAME_RE = re.compile(r"^[A-Za-zÀ-ÖØ-öø-ÿ]+(?:[ '\u2019-][A-Za-zÀ-ÖØ-öø-ÿ]+)*$")
POSTAL_CODE_RE = re.compile(r"^\d{4}-\d{3}$")
PASSWORD_RE = re.compile(r"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,128}$")


def clean_text(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    cleaned = value.strip()
    return cleaned or None


def validate_email(value: str) -> str:
    email = value.strip().lower()
    if not EMAIL_RE.fullmatch(email):
        raise ValueError("Invalid email address.")
    domain = email.rsplit("@", 1)[-1]
    if domain in DISPOSABLE_EMAIL_DOMAINS:
        raise ValueError("Invalid email address.")
    return email


def validate_password(value: str) -> str:
    if not PASSWORD_RE.fullmatch(value):
        raise ValueError("Password must contain uppercase and lowercase letters, a number, and a special character.")
    return value


def validate_name(value: Optional[str]) -> Optional[str]:
    name = clean_text(value)
    if name is None:
        return None
    if len(name) < 2:
        raise ValueError("Name must contain at least 2 characters.")
    if len(name) > 100:
        raise ValueError("Name must contain at most 100 characters.")
    if name.isdigit():
        raise ValueError("Name cannot contain only numbers.")
    if not NAME_RE.fullmatch(name) or not any(char.isalpha() for char in name):
        raise ValueError("Enter a valid full name.")
    return name


def normalize_phone(value: Optional[str]) -> Optional[str]:
    phone = value.strip().replace(" ", "") if value else ""
    if not phone:
        return None
    if phone.startswith("+351"):
        national = phone[4:]
        normalized = phone
    else:
        national = phone
        normalized = phone
    if not national.isdigit():
        raise ValueError("Phone number must contain only digits.")
    if len(national) != 9 or not national.startswith("9"):
        raise ValueError("Invalid Portuguese phone number.")
    return normalized


def validate_portuguese_tax_id(value: Optional[str]) -> Optional[str]:
    tax_id = value.strip() if value else ""
    if not tax_id:
        return None
    if not tax_id.isdigit() or len(tax_id) != 9:
        raise ValueError("Portuguese tax ID must contain exactly 9 digits.")
    checksum = sum(int(tax_id[index]) * (9 - index) for index in range(8))
    check_digit = 11 - (checksum % 11)
    if check_digit >= 10:
        check_digit = 0
    if check_digit != int(tax_id[-1]):
        raise ValueError("Invalid Portuguese tax ID.")
    return tax_id


def validate_postal_code(value: Optional[str]) -> Optional[str]:
    postal_code = value.strip() if value else ""
    if not postal_code:
        return None
    if not POSTAL_CODE_RE.fullmatch(postal_code):
        raise ValueError("Postal code must use the Portuguese XXXX-XXX format.")
    return postal_code

File: frontend/src/utils/validation.ts
export type FieldErrors<T extends string> = Partial<Record<T, string>>

const disposableEmailDomains = new Set([
  "10minutemail.com",
  "guerrillamail.com",
  "mailinator.com",
  "tempmail.com",
  "temp-mail.org",
  "yopmail.com",
])

const emailPattern = /^[^\s@<>]+@[^\s@<>]+\.[^\s@<>]{2,}$/
const namePattern = /^[A-Za-zÀ-ÖØ-öø-ÿ]+(?:[ '\u2019-][A-Za-zÀ-ÖØ-öø-ÿ]+)*$/
const postalCodePattern = /^\d{4}-\d{3}$/
const passwordPattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,128}$/

export function normalizeText(value: string) {
  return value.trim()
}

export function normalizePhone(value: string) {
  return value.replace(/\s+/g, "")
}

export function validateEmail(value: string) {
  const email = normalizeText(value).toLowerCase()
  if (!emailPattern.test(email)) return i18n.t("validation.emailInvalid", { ns: "errors" })
  const domain = email.split("@").pop() ?? ""
  if (disposableEmailDomains.has(domain)) return i18n.t("validation.emailInvalid", { ns: "errors" })
  return ""
}

export function validatePassword(value: string) {
  return passwordPattern.test(value)
    ? ""
    : i18n.t("validation.passwordStrength", { ns: "errors" })
}

export function validateName(value: string, required = true) {
  const name = normalizeText(value)
  if (!name) return required ? i18n.t("validation.nameInvalid", { ns: "errors" }) : ""
  if (name.length < 2) return i18n.t("validation.nameTooShort", { ns: "errors" })
  if (name.length > 100) return i18n.t("validation.nameTooLong", { ns: "errors" })
  if (/^\d+$/.test(name)) return i18n.t("validation.nameOnlyNumbers", { ns: "errors" })
  if (!namePattern.test(name) || !/[A-Za-zÀ-ÖØ-öø-ÿ]/.test(name)) {
    return i18n.t("validation.nameInvalid", { ns: "errors" })
  }
  return ""
}

export function validatePhone(value: string, required = true) {
  const phone = normalizePhone(value)
  if (!phone) return required ? i18n.t("validation.phoneInvalid", { ns: "errors" }) : ""
  const national = phone.startsWith("+351") ? phone.slice(4) : phone
  if (!/^\+?\d+$/.test(phone) || (phone.includes("+") && !phone.startsWith("+351"))) {
    return i18n.t("validation.phoneDigits", { ns: "errors" })
  }
  if (!/^\d{9}$/.test(national) || !national.startsWith("9")) {
    return i18n.t("validation.phoneInvalid", { ns: "errors" })
  }
  return ""
}

export function validateNif(value: string, required = false) {
  const taxId = normalizeText(value)
  if (!taxId) return required ? i18n.t("validation.taxIdLength", { ns: "errors" }) : ""
  if (!/^\d{9}$/.test(taxId)) return i18n.t("validation.taxIdLength", { ns: "errors" })
  const checksum = taxId
    .slice(0, 8)
    .split("")
    .reduce((sum, digit, index) => sum + Number(digit) * (9 - index), 0)
  let checkDigit = 11 - (checksum % 11)
  if (checkDigit >= 10) checkDigit = 0
  return checkDigit === Number(taxId[8]) ? "" : i18n.t("validation.taxIdInvalid", { ns: "errors" })
}

export function validatePostalCode(value: string, required = false) {
  const postalCode = normalizeText(value)
  if (!postalCode) return required ? i18n.t("validation.postalCodeInvalid", { ns: "errors" }) : ""
  return postalCodePattern.test(postalCode) ? "" : i18n.t("validation.postalCodeInvalid", { ns: "errors" })
}
import i18n from "../i18n"
